persistence/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedActorOf.scala

```diff
@@ -0,0 +1,161 @@
+    persistence: EventSourcedPersistence[F]
+  ): Actor = ActorOf[F] { actorCtx =>
+    for {
+      eventSourced    <- eventSourcedOf(actorCtx).toResource
+      recoveryStarted <- eventSourced.value
+
+      snapshotStore <- persistence.snapshotStore[S](eventSourced).toResource
+      eventStore    <- persistence.eventStore[E](eventSourced).toResource
+
+      snapshot <- snapshotStore.latest.toResource
+
+      recovering <- recoveryStarted(
+        snapshot.map(_.metadata.seqNr).getOrElse(SeqNr.Min),
+        snapshot.map(_.asOffer)
+      )
+
+      seqNr <- recovering.replay.use { replay =>
+        val snapSeqNr = snapshot.map(_.metadata.seqNr).getOrElse(SeqNr.Min)
+        val fromSeqNr = snapshot.map(_.metadata.seqNr + 1).getOrElse(SeqNr.Min)
+        for {
+          events <- eventStore.events(fromSeqNr)
+          seqNrL <- events.foldWhileM(snapSeqNr) {
+            case (_, EventStore.Event(event, seqNr)) => replay(event, seqNr).as(seqNr.asLeft[Unit])
+            case (_, EventStore.HighestSeqNr(seqNr)) => seqNr.asLeft[Unit].pure[F]
+          }
+          seqNr <- seqNrL match {
+            case Left(seqNr) => seqNr.pure[F]
+            case Right(_)    =>
+              // function used in events.foldWhileM always returns Left
+              // and sstream.Stream.foldWhileM returns Right only if passed function do so
+              // thus makes getting Right here impossible
+              new IllegalStateException("should never happened").raiseError[F, SeqNr]
+          }
+        } yield seqNr
+      }.toResource
+
+      currentSeqNr <- Ref[F].of(seqNr).toResource
+
+      snapshotter = new Snapshotter[F, S] {
+
+        def save(seqNr: SeqNr, snapshot: S): F[F[Instant]] = snapshotStore.save(seqNr, snapshot)
+
+        def delete(seqNr: SeqNr): F[F[Unit]] = snapshotStore.delete(seqNr)
+
+        def delete(criteria: SnapshotSelectionCriteria): F[F[Unit]] = snapshotStore.delete(criteria.asStoreCriteria)
+
+      }
+
+      journaller = new Journaller[F, E] {
+
+        val append = new Append[F, E] {
+
+          def apply(events: Events[E]): F[F[SeqNr]] =
+            currentSeqNr
+              .modify { seqNr0 =>
+                val seqNr1 = seqNr0 + events.size
+                seqNr1 -> seqNr0
+              }
+              .map { seqNr0 =>
+                var seqNr = seqNr0
+                def nextSeqNr: SeqNr = {
+                  seqNr = seqNr + 1
+                  seqNr
+                }
+                events.map(e => EventStore.Event(e, nextSeqNr))
+              }
+              .flatMap { events =>
+                eventStore.save(events)
+              }
+
+        }
+
+        val deleteTo = new DeleteEventsTo[F] {
+
+          def apply(seqNr: SeqNr): F[F[Unit]] = eventStore.deleteTo(seqNr)
+
+        }
+
+      }
+
+      receive <- recovering.completed(seqNr, journaller, snapshotter)
+    } yield receive.contramapM[Envelope[Any]](_.cast[F, C])
```

Comment:
Technically, it will throw `ClassCastException(s"${ self.getClass.getName } cannot be cast to ${ tag.runtimeClass.getName }")` if an actor receives a command, which is not `C`.

Could it happen in practice? Is it intended behavior? I do not know if it is, so if it is, then feel free to ignore.

If it is possible to happen then, may be, consider, more informative message, so the developer know where to look. I.e. "actor X received the wrong message Y" or something like that. :thinking:

Reply:
https://github.com/evolution-gaming/akka-effect/pull/276#discussion_r1450321368